Add algorithmic backend

f52985 opened the pull request:
This pull request adds algorithmic backends (prose, interpreter):
* `src/al` directory contains the syntax and other utility functions for the intermediate representation algorithmic language AL.
* `src/backend-prose`, `src/backend-interpreter` contains the code for generating prose and (meta-level) interpreter from the input spec.
* Targets `Prose` and `Interpreter` can be selected as the target by command argument.
* `test-prose` and `test-interpreter` contains the test code for these backends.

It also contains a few changes to other part of the code:
* `spec/wasm-3.0` now contains syntax, validation rules, and reduction rules for SIMD instructions. (which is planned to be copied to `spec/wasm-2.0` after testing with official test-suite)
* The splicing is now done via the separated target `Splice` instead of `Latex`, and splicing now creates a new file instead of modifying the input file in-place.
* There's a new print function, `structured_string_of` for `el` and `il`, which prints the AST in a "structured" manner, mainly used during debugging.

rossberg commented on spectec/src/util/lib.ml:
```diff
@@ -11,6 +11,14 @@ struct
+  let split_last_opt = function
+    | x::[] -> [], Some x
+    | x::xs -> let ys, y = split_last xs in x::ys, Some y
+    | [] -> [], None
```
Nit: to avoid code duplication, perhaps reimplement split_last in terms of split_last_opt now?

rossberg commented on spectec/src/util/record.ml:
```diff
@@ -0,0 +1,49 @@
+
+(* Helper functions for record *)
+module Record = struct include List
```
Why this nested module?

f52985 replied:
Good point, that's because of inexperience in using Ocaml module system 😅.

rossberg commented on spectec/src/util/record.ml:
```diff
@@ -0,0 +1,49 @@
+
+(* Helper functions for record *)
+module Record = struct include List
+  (* Record implements finite mapping *)
```
Not sure how heavily this is used, but it seems rather inefficient. Wouldn't you rather want to use the OCaml lib's Map module for this purpose?

f52985 replied:
That's because we wanted to keep the order of keys.

`Record` is mainly used for expressing the key-value structure in the spec such as
```
syntax funcinst hint(desc "function instance") =
  { TYPE deftype,
    MODULE moduleinst,
    CODE func }
```
and we wanted to keep this order when printing prose.

I must admit though, on second thought, there are some places where `Record` is used even if order of keys is not important. For those parts, we will refactor into using Map.

rossberg replied:
I see, that makes sense. Yes, I was wondering about other uses, because the simple record use case wouldn't need most of the generic traversal functions in this module.

rossberg commented on spectec/src/el/free.ml:
```diff
@@ -115,7 +114,7 @@ and free_exp e =
   | StrE efs -> free_nl_list free_expfield efs
   | CallE (id, args) -> union (free_defid id) (free_list free_arg args)
-  | IterE (e1, iter) -> union (free_exp e1) (free_iter iter)
+  | IterE (e1, iter) -> union (diff (free_exp e1) (binding_iter iter)) (free_iter iter)
```
The semantics I had in mind so far was one where
```
(...)^(i<n)
```
effectively is a shorthand for
```
(...)^n /\ i^n = 0 ... (n-1)
```
i.e., `i` is simply a sequence of length n, with nothing else special about it. The advantage is that this avoids the introduction of local binders, which we do not currently have, and which might require more thought.

Was there a particular reason you had to change this interpretation? If not, it would probably make sense to split this change from this PR, so that we can discuss it separately.

f52985 replied:
The fundamental reason was to fix a false type error.

For example, the following `sample.watsup` file
```
relation R: nat* -> nat*

var n: nat
var m: nat

rule R/r: n* -> m^8
  -- (if m = n*[i])^(i<8)
```
results in an unwanted type error:
```
$ ./watsup sample.watsup
sample.watsup:6.1-7.26: type error: rule contains unbound variable(s) `i`
```

which prevented writing reduction rules for one of SIMD instructions. (Specificaly, `VLOAD`)

After seeing this message, my first impression was: "the variable `i` should not be considered free in the entire `IterE` expression," and that's the reason for changing the implementation in a way that `i` is excluded from free variables in the entire `IterE` expression.

Since that was the only reason, I'm happy with any interpretation, as long as the example above is considered well-typed.

rossberg replied:
Ah, okay. I'll think about it a little and tweak main to make that example work.

rossberg replied:
This actually was intended to work, but there was a bug in the computation of bound variables, see #54.

f52985 commented:
Thanks for the review! We'll soon address each issue. Here are the thoughts on the review in the main comments.

> The one structural change I'd request is to remove the duplicate of the reference_interpreter inside backend-interpreter. Obviously, we do not want to maintain two copies of this code. The original code is in the same repo, so it should be possible to set things up such that we can import it from there.

Yes, that was what we also wanted to do, but we couldn't figure out how to do. Although they are in the same repository, they are separated dune-project, and we could not find a way to import reference interpreter in spectec.

However, there is another problem: `reference_interpreter` should be able to parse wasm3 proposals in order to parse wasm3 tests. (In fact, current `reference_interpreter` is actually copied from gc proposal's repo, not this repo). Even if we (somehow figure out how to) use interpreter in the same repo, it won't be able to parse wasm3 tests.

> One other change I'm not sure I like is the way how splice output files are given. While I sympathise with the desire to avoid destructive file update, a separate list of file names is at odds with Unix conventions. I was following the sed model, which is easier to use directly – e.g., I envisioned that one can do something as simple as cp -R src out; watsup src/**/*.watsup to transform an entire document tree. In contrast, the new command syntax essentially necessitates a Makefile to produce all the output file paths economically.
If we wanted a non-mutable model, then the Unix way would be to derive output file names automatically from input names, e.g., by appending an extension. Or allow only one file at a time. Not sure either is better, though, when the goal is to run over entire directory trees.

I see, so the main point here is that we do need in-place file mutation, especially because we want to mutate the entire directory. In that case, I would like to suggest to add a command option that flags in-place mutation. That is what `sed` is also doing. `sed` prints the result to the stdout by default and do the mutation when `-i` option is given, so we can take the similar approach.

rossberg reviewed (commented):
> However, there is another problem: `reference_interpreter` should be able to parse wasm3 proposals in order to parse wasm3 tests. (In fact, current `reference_interpreter` is actually copied from gc proposal's repo, not this repo). Even if we (somehow figure out how to) use interpreter in the same repo, it won't be able to parse wasm3 tests.

Ah, good point. In that case, instead of checking in the copy, can we add something to our Makefile that checks out the GC proposal repo in place?

> I see, so the main point here is that we do need in-place file mutation, especially because we want to mutate the entire directory. In that case, I would like to suggest to add a command option that flags in-place mutation. That is what `sed` is also doing. `sed` prints the result to the stdout by default and do the mutation when `-i` option is given, so we can take the similar approach.

Yes, that's a great idea.

f52985 commented on spectec/spec/wasm-3.0/1-syntax.watsup:
```diff
@@ -262,7 +320,8 @@ var mo : memop
 syntax c = nat
 syntax c_numtype = nat  ;; TODO
-syntax c_vectype = nat  ;; TODO
+syntax c_vectype = byte*;; TODO
```
Our first attempt of using `nat` for `vec_type` was not successful (mainly due to the fact that in algorithmic backends, nat is encoded to `Int64` in Ocaml, which can not store 128 bit numbers), and using `byte*` was a temporary hack (thus TODO).

Considering that we ultimately want to use `i128`, we will try again to see if using nat would be possible.

rossberg replied:
Oh, I see, that's a problem all right. :) Since i128 will be a refinement of nat, that will remain so. We should probably change the implementation of nat and int to use the zarith library (https://github.com/ocaml/Zarith).

f52985 replied:
It turns out that keeping `c_vectype` as `nat` in the spec itself was not that difficult, after all.

By using a bit of hack during translation, algorithmic backends still use bytes sequence (string) as underlying representation for 128bit values, just as reference interpreter does.

f52985 commented on spectec/src/il/print.ml:
```diff
@@ -282,3 +282,350 @@ let rec string_of_def d =
+
+let sprintf = Printf.sprintf
+
+let structured_string_of_list stringifier = function
```
Oh we were not aware of the existence of `util/sexpr`. That would make things much simpler! We'll have a look at it.

f52985 commented:
Most comments are now resolved, and there are 2 remaining:

1. Pretty printer using `sexpr`
Considering the effort to convert each AST into sexpr, I think the pretty printer should be merged with a separated pull request.

2. Copying reference interpreter
I tried to do some workaround for using reference interpreter in gc proposal without copying it, but it makes the compiling process rather clumsy. Considering the fact that
- the sole reason for using reference interpreter is its parser,
- the changes in the actual reference interpreter need not be reflected to the copied reference interpreter for the purpose of parsing,
- and the parser would also be (hopefully) automatically generated using SpecTec,
I think using the fixed, copied version of reference interpreter would be OK for now.

rossberg commented on spectec/src/frontend/elab.ml:
```diff
@@ -1474,7 +1474,7 @@ let elab_def env d : Il.def list =
     let binds' = make_binds env' free dims d.at in
     let rule' = Il.RuleD (id2, binds', mixop, tup_exp' es' e.at, prems') $ d.at in
-    env.rels <- rebind "relation" env.rels id1 (t, rules' @ [rule']);
+    env.rels <- rebind "relation" env.rels id1 (t, rules' @ [rule'] );
```
Spurious change?

rossberg commented on spectec/src/backend-latex/config.ml:
```diff
@@ -23,26 +12,12 @@ type config =
-      {token = "$$"; prefix = ".. math::\n   "; suffix = ""; indent = "   "};
-    ]
-  }
+let sphinx = default
```
The latex/sphinx shorthands are no longer meaningful in this module, we can remove them.

rossberg commented on spectec/src/backend-splice/config.ml:
```diff
@@ -0,0 +1,43 @@
+    (* Prose *)
+    prose : Backend_prose.Config.config;
+  }
+
```
Let's add the `type t = config` shorthand here as well.

rossberg commented on spectec/src/backend-latex/render.mli:
```diff
@@ -5,17 +5,24 @@ open El.Ast
 type env
 
-val env : Config.t -> El.Ast.script -> env
+val env : Config.config -> El.Ast.script -> env
```
```suggestion
val env : Config.t -> El.Ast.script -> env
```

rossberg commented on spectec/src/backend-latex/render.mli:
```diff
@@ -5,17 +5,24 @@ open El.Ast
 
-val env : Config.t -> El.Ast.script -> env
+val env : Config.config -> El.Ast.script -> env
+val new_env : Config.config -> env
```
Is this needed?

jaehyun1ee replied:
Just some clutter that I forgot to remove, thanks!

rossberg commented on spectec/src/backend-splice/splice.ml:
```diff
@@ -271,6 +319,29 @@ let splice_anchor env src anchor buf =
   ignore (
     try_exp_anchor env src r ||
+(*
+<<<<<<< HEAD:spectec/src/backend-splice/splice.ml
```
Should this be cleaned up?

rossberg commented on spectec/src/backend-splice/splice.mli:
```diff
@@ -0,0 +1,10 @@
+
+val env : Config.config -> string list -> string list -> El.Ast.script -> Backend_prose.Prose.prose -> env
+
+val get_render_prose : env -> Backend_prose.Render.env
```
Is this needed?

jaehyun1ee replied:
Also unnecessary, removed.

rossberg commented on spectec/src/il/validation.ml:
```diff
@@ -526,7 +526,7 @@ and valid_prem env prem =
   | IfPr e ->
     valid_exp env e (BoolT $ e.at)
-  | LetPr (e1, e2) ->
+  | LetPr (e1, e2, _ids) ->
```
What is the semantics of these ids? Shouldn't they be checked here?

f52985 replied:
`ids` are there for partial bindings.

For example, if the premise `-- where (a, b) = c`, is a binding for `a`, but not for `b`, then the `ids` will be the list `["a"]`.

I added a check if `ids` is a subset of the free variables in the left-hand side expression `e1`.

rossberg commented on spectec/README.md:
```diff
@@ -233,21 +233,18 @@ Then, Sphinx builds the `rst` files into desired formats such as pdf or html.
 $ git checkout al
 $ make
-$ ./watsup spec/* --animate --sideconditions --interpreter
+$ ./watsup spec/* --animate --sideconditions --interpreter test-interpreter/sample.wast
```
Actually remove the redundant flags? ;)

f52985 replied:
Oops, my bad. They are surely removed this time.

rossberg reviewed (commented):
Excellent, thanks!

Re 1: Yes, that's totally fine with me.

Re 2: Does just git cloning it into a subdirectory not work? That could be performed by something as simple as an explicit setup target in the Makefile (though doing it automatically would of course be nicer). I'd really like to avoid having copied code, as that always becomes a liability quickly.

f52985 commented:
All comments are resolved now.

> Re 2: Does just git cloning it into a subdirectory not work? That could be performed by something as simple as an explicit setup target in the Makefile (though doing it automatically would of course be nicer). I'd really like to avoid having copied code, as that always becomes a liability quickly.

I modified the `Makefile` in a way that it clones the repository for gc-proposal before building with a fixed commit hash (so that it can prevent the situation where new commits to gc-proposal's interpreter break the compilation). This cloned repository is git-ignored, and removed upon `make clean`.

rossberg commented on spectec/src/il/validation.ml:
```diff
@@ -526,8 +526,14 @@ and valid_prem env prem =
+    let lhs_ids = (Free.free_exp e1).varid in
+    if not (Free.Set.subset target_ids lhs_ids) then
+      error prem.at ("binding premise's target(s) `" ^ String.concat "," ids ^
+        "` is not contained in the left hand side expression `" ^ string_of_exp e1)
```
Nit: expressions can become quite large, so I wouldn't include them in single line error messages.
```suggestion
        "` is not contained in the left-hand side expression")
```

rossberg reviewed (approved):
Excellent, thanks!
